PR for Code Review of Line Itemizer Subtotal Calculator Feature

timschafli opened the pull request:
Updated Line Itemizer with additional fields for controlling a new Subtotal Calculator feature.

You can see this in Action here: https://zapier.com/app/editor/55581116/overview
And here: https://zappy.zapier.com/0E0F28BC-6B07-44FF-BED9-A904230D5F3C.mp4
And here: https://zapier.slack.com/archives/CDEL5KD40/p1554268618005600

robwa10 commented:
@timschafli this is a neat little addition to this feature, definitely can see how this would be useful! I've picked out one or two minor things in the code but I'll leave that to @kgodtfredsen and @cvanderkolk 
 since they've got more context.  

Looking at the UI I'd suggest reordering the layout a little and some changes to the Help Text. For the reordering I'd move the choice to use the subtotal above all the settings. I wonder if users will see all the fields and try to set them, then find out they didn't need to.  

I'm unsure if you can 'hide' additional fields based on a selection in another field. But if we can make all the subtotal settings appear when the dropdown is set to "yes" that would be the ideal. I know it can be done in [CLI](https://github.com/zapier/zapier-platform-cli#user-content-customdynamic-fields) via `altersDynamicFields: true` but I don't know about Python apps.  

Great stuff going on here!

robwa10 commented on transformer/transforms/util/line_itemizer.py:
```diff
@@ -107,10 +156,59 @@ def fields(self, *args, **kwargs):
+            {
+                "key": "subtotal_help",
+                "type": "copy",
+                "help_text": "If you have 'Price' and 'Quantity' properties in your line items, Line Itemizer "
```
I can see this getting a little confusing for users. What about something like this?  
"If you have price and quantity properties in your line items, Line Itemizer will try to multiply those values together to create a subtotal property. [Learn more about the Subtotal property here](https://zapier.com/help/formatter/#create-your-own-line-items-for-an-invoicing-action-using-the-line-itemizer-utility)."  

And then 'force' the user to set the Price and Quantity property in the input fields.

robwa10 commented on transformer/transforms/util/line_itemizer.py:
```diff
@@ -107,10 +156,59 @@ def fields(self, *args, **kwargs):
+            {
+                "key": "subtotal_help",
+                "type": "copy",
+                "help_text": "If you have 'Price' and 'Quantity' properties in your line items, Line Itemizer "
```
I can see this getting a little confusing for users. What about something like this?
"If you have price and quantity properties in your line items, Line Itemizer will try to multiply those values together to create a subtotal property. Learn more about the Subtotal property here."

And then 'force' the user to set the Price and Quantity property in the input fields.

timschafli replied:
I like having the default values that can be overwridden (even though it feels like a lot of fields) because if someone had this in an existing Zap and their values were even in the wrong case, they would have to remap their whole Zap to use this feature. If you can specify which fields to use, you can get around that issue. I wanted to keep the spirit of being able to name every property as the user sees fit.

robwa10 commented on transformer/transforms/util/line_itemizer.py:
```diff
@@ -107,10 +156,59 @@ def fields(self, *args, **kwargs):
+                "key": "my_subtotal_name",
+                "label": "Subtotal Property Name",
+                "default": "Subtotal",
+                "help_text": "Set the Subtotal Property Name here. Default is 'Subtotal'. Formatter will not "
```
Would it be better to force the user to set a property name here as well and tell them it has to be unique from any other name already set?

timschafli replied:
Do you mean a required field? The reason I didn't go with that was to better accommodate users who don't need to use the feature. With the new defaults I'll push to the PR shortly if they ignore all the subtotal fields and nothing will happen. I think it should be straightforward that you can't have two fields with the same name, and that if you do, the one that you specified first in the Action will be the one that is used.

robwa10 commented on transformer/transforms/util/line_itemizer.py:
```diff
@@ -107,10 +156,59 @@ def fields(self, *args, **kwargs):
+                "label": "Create Subtotal Property?",
+                "default": "Yes",
+                "choices": "Yes,No",
+                "help_text": "Default is 'Yes'. Set to 'No' if you don't want to create this property.",
```
Is the use case with Line-Itemizer such that 'Yes' should be default? If the input fields can be dynamically generated based on this field setting I think 'No' may be the better choice.

timschafli replied:
Initially I figured most folks would want this, and I took precautions not to break existing Line Itemizers that had a Subtotal property already. However, on further reflection, I changed the default to No, so I was in total agreement already when I read this :)

kgodtfredsen replied:
I agree on this, No seems like a better default.

robwa10 commented on transformer/transforms/util/line_itemizer.py:
```diff
@@ -107,10 +156,59 @@ def fields(self, *args, **kwargs):
+                "key": "my_subtotal_toggle",
+                "label": "Create Subtotal Property?",
+                "default": "Yes",
+                "choices": "Yes,No",
```
I _think_ there's supposed to be a space after the comma re:style guide. `"Yes,No"` -> `"Yes, No"`. But I'm not sure how strictly we follow [Pep8](https://www.python.org/dev/peps/pep-0008/#whitespace-in-expressions-and-statements)

timschafli replied:
In the context of Web Builder apps, the legacy docs have it both ways, but more examples show  no space after the comma: https://zapier.com/developer/documentation/v2/reference/#static-dropdowns

timschafli commented:
@robwa10 Thank for taking a look! Totally agree on moving the Toggle up to the top. Would love to show/hide fields and I did spend some time trying to do it, but because there is already dynamic field loading going on with Transform selection, it proved too difficult for me to figure out, but I'd love that, too. Another reason to go CLI / Godzilla. I did make some changes to the help text already, and about to update this PR with that. Will review again soon with your input in mind!

timschafli commented:
Just updated this PR with another Commit that includes changes to the help text, order of the Subtotal fields, and some of the defaults. 

Also added another property generator suggested by Nick for creating user-accessible numbering for line-items as well. Hope it's not too much. It really seemed to be at home in the Line Itemizer and was super quick to create.

kgodtfredsen commented:
Hey Tim, I'm going to do a concept review here first - I like the change to move the default to No and to the Top, but I'm a bit worried about the complexity of the transform. We don't have any other Transforms that take on this level of functionality, and I worry we are getting to the point where it's almost like an app. In general, Formatter transforms do one thing today, and the code is relatively simple can be picked up by anybody and moved forward. I'd rather not get to a situation where we have 30 transforms that are like this, then one that is not. Its a larger question on how we should deal with this over time  - as we do need utilities that have more functionality. Hopefully, we can start talking about this sooner than later.

In this vein, could the subtotals transform live on its own and have more functionality? If I have a Shopify Trigger where I'm already getting LIs, but want some calc of fields, where would that fit in here (that is the calc scenario that we have in help today). 

In this vein, I'd also remove the numbering property. If we have a Loki with 10 folks asking for this, then we could create another transform for doing this - Line Itemizer Prettifier? (blech!). Lots of things would work great inside Line-Itemizer today, but that is true for many other transforms too, and we've seemingly resisted that over time.

timschafli commented on transformer/transforms/util/line_itemizer.py:
```diff
@@ -107,10 +173,81 @@ def fields(self, *args, **kwargs):
+                "[Learn more about the Subtotal property here](https://zapier.com/help/formatter/#create-line-items-with-subtotals-using-the-line-itemizer-utility).",
+            },
+            {
+                "type": "unicode",
```
I've gone ahead and changed this to a Bool that defaults to false. I also changed it so that it will overwrite an existing Subtotal field to cut down on help text because that seems more obvious to me. 
In some cases, users may want to specify their own Subtotal that they created using a code Step or a Spreadsheet Style Formula. The Subtotal feature takes care of 99% of those cases, but it's possible someone might want to do their own thing.

timschafli commented on transformer/transforms/util/line_itemizer.py:
```diff
@@ -107,10 +173,81 @@ def fields(self, *args, **kwargs):
+            },
+            {
+                "type": "unicode",
+                "required": False,
```
Done, agreed

timschafli commented on transformer/transforms/util/line_itemizer.py:
```diff
@@ -107,10 +173,81 @@ def fields(self, *args, **kwargs):
+                "Default is '2'.",
+            },
+            {
+                "key": "numbering_help",
```
I think this should be left in. 
I added the Decimal library to increase the accuracy of the calculation generally.
Also, I've had personal experience already of cases where specifying 5 decimal places would be necessary for QBO to get taxes right, otherwise you fall into the "off by a penny" issues.

This comes into play when the quantity is not a whole number and the user has to deal with exciting tax codes like where I live :P 
For example, imagine someone billing for 2.75 hours of work at 69.99 per hour, then applying two separate tax rates of 5% and 7%. The extra resolution can potentially resolve those errors. Most folks probably won't touch this dial, but the ones who do will be glad it's there :)

timschafli commented on transformer/transforms/util/line_itemizer.py:
```diff
@@ -93,9 +132,36 @@ def transform(self, input_key, my_dict={}, **kwargs):
                 except IndexError as e:
                     pass
+            if (
+                my_price in this_line_item.keys()
```
Fixed!

timschafli commented on transformer/transforms/util/line_itemizer.py:
```diff
@@ -107,10 +173,81 @@ def fields(self, *args, **kwargs):
+            {
+                "type": "unicode",
+                "required": False,
+                "key": "my_price",
```
Done!

cvanderkolk commented on transformer/transforms/util/line_itemizer.py:
```diff
@@ -31,34 +33,42 @@ def build_input_field(self):
 
-    def transform(self, input_key, my_dict={}, **kwargs):
+    def transform(
+        self, input_key, my_dict={}, decimals="2", subtotal_toggle=False, **kwargs
```
I would think that `decimals` would default to `2` as an integer instead of as a string here. Is there a particular reason why this is a string here?

cvanderkolk replied:
Also, it can be problematic to use a mutable value as a default argument like you're doing here with `my_dict`: https://docs.quantifiedcode.com/python-anti-patterns/correctness/mutable_default_value_as_argument.html#best-practice

 Best way to handle this is to initialize it with `None` here and then set its value below:
```
if my_dict is None:
  my_dict = {}
```

timschafli replied:
Made those changes. No particular reason for it to be a string. I think I was just used to other inputs coming in as Strings. I'll pay closer attention to my field value types.

cvanderkolk commented on transformer/transforms/util/line_itemizer.py:
```diff
@@ -71,6 +81,31 @@ def transform(self, input_key, my_dict={}, **kwargs):
         # initialize output and other variables
         output = {input_key: []}
         longest_array = 0
+        subtotal = "Subtotal"
```
I know I didn't catch this on our last review round, but ideally all these dict property names would look a little more "pythonic", that is to say `subtotal` instead of `Subtotal` and `line_items` instead of `Line Items`. You don't have to change the old keys to preserve backward compatibility, but moving forward I recommend you stick to Python naming conventions: https://www.python.org/dev/peps/pep-0008/#descriptive-naming-styles

timschafli replied:
Made this change. I wonder, would it be possible to set the input value for anyone who doesn't have it set to "Line Item(s)" so we don't break their Zaps so that we can also change the value in the transformer service to "line_items"?

cvanderkolk commented on transformer/transforms/util/line_itemizer.py:
```diff
@@ -71,6 +81,31 @@ def transform(self, input_key, my_dict={}, **kwargs):
+                elif k.lower() == "qty":
+                    qty = k
+
+        decimal_price = Decimal(0)
```
I don't think you need to initialize these here since you define them below. There might be a nuance I'm missing here, though...

timschafli replied:
Tried without initializing them up there and it seemed fine. I probably needed it before for something, but not anymore.

cvanderkolk commented on transformer/util.py:
```diff
@@ -96,15 +96,17 @@ def try_parse_date(date_value, from_format=None):
     try:
-        if from_format:
+        # Arrow does not handle negative Unix Timestamps (pre-1970) when explicitly using the token 'X'
+        if from_format and not from_format == 'X':  
```
Does this have anything to do with the Line Itemizer feature? If not, I would save this for a different PR to maintain separation of concerns—you generally only want to ship a feature at a time here if we can prevent it.I know it's tempting to sneak small fixes into a PR you already have open, but doing that can lead to confusion when we ship if a bug arises.

timschafli replied:
This does not, and was not intended to be committed here. I've started a slack chat with you in #wg-formatter to figure out how best to remove this.

timschafli replied:
I found and removed the commit from this branch.

cvanderkolk commented on transformer/transforms/util/line_itemizer.py:
```diff
@@ -93,9 +132,36 @@ def transform(self, input_key, my_dict={}, **kwargs):
+                try:
+                    price = Decimal(this_line_item[my_price])
+                    qty = Decimal(this_line_item[my_qty])
+                    this_line_item.update(
```
@timschafli have you tried doing my recommendation here? I think that's a bit more straightforward than `.update`.  Even if we're using a different name for the property, you could do `this_line_item[subtotal]` to use the variable.

timschafli replied:
Sorry, missed this the first time around. Implemented your change now. Still requires to convert the result to a String (preferred) or Float first otherwise the output is something like "Decimal(1.00)". I do think assigning a value to the key is more legible this way.
